File: amr/localization/sensor_model.py
"""Likelihood-field range-finder sensor model.

Precomputes a distance field (metres to the nearest occupied cell) over the
map. For each particle, the valid beam endpoints are transformed into the world
frame, looked up in the distance field, and scored with the per-beam likelihood
``q = z_hit * exp(-d^2 / (2 sigma^2)) + z_rand / range_max``. The particle
weight is ``exp(sum_b ln q)``. Fully vectorized over particles and beams.
See Probabilistic Robotics (Thrun et al.) Table 6.3.
"""
from __future__ import annotations

import numpy as np

from amr.core.config import LikelihoodConfig
from amr.core.geometry import distance_field
from amr.core.types import LaserScan, OccupancyGrid


class LikelihoodField:
    def __init__(self, grid: OccupancyGrid, cfg: LikelihoodConfig):
        self.grid = grid
        self.cfg = cfg
        occupied = grid.data >= 65
        self.field = distance_field(occupied, grid.resolution, cfg.max_dist)
        # Cells that are unknown in the source map are treated as max_dist so
        # they contribute no informative hit likelihood.
        self.unknown = grid.data < 0

    def weigh(self, particles: np.ndarray, scan: LaserScan) -> np.ndarray:
        """(N,) unnormalized weights."""
        particles = np.asarray(particles, dtype=float)
        n = particles.shape[0]
        cfg = self.cfg
        grid = self.grid
        res = grid.resolution
        max_dist = cfg.max_dist
        sigma = cfg.sigma_hit

        valid = scan.valid_mask()
        sub = max(int(cfg.beam_subsample), 1)
        angles = scan.angles()
        ranges = scan.ranges

        a = angles[valid][::sub]
        r = ranges[valid][::sub]

        if a.size == 0:
            return np.ones(n, dtype=float)

        # Beam endpoints in the robot frame: (B, 2).
        bx = r * np.cos(a)
        by = r * np.sin(a)

        th = particles[:, 2]                       # (N,)
        cos_t = np.cos(th)[:, None]                # (N, 1)
        sin_t = np.sin(th)[:, None]
        px = particles[:, 0][:, None]              # (N, 1)
        py = particles[:, 1][:, None]

        # Broadcast-transform every beam endpoint by every particle: (N, B).
        wx = px + cos_t * bx[None, :] - sin_t * by[None, :]
        wy = py + sin_t * bx[None, :] + cos_t * by[None, :]

        cols = np.floor((wx - grid.origin_x) / res).astype(np.int64)
        rows = np.floor((wy - grid.origin_y) / res).astype(np.int64)

        in_bounds = ((rows >= 0) & (rows < grid.rows) &
                     (cols >= 0) & (cols < grid.cols))

        rc = np.clip(rows, 0, grid.rows - 1)
        cc = np.clip(cols, 0, grid.cols - 1)

        d = self.field[rc, cc].astype(float)       # (N, B)
        unknown = self.unknown[rc, cc]
        # Out-of-bounds or unknown cells contribute the max distance.
        d = np.where(in_bounds & ~unknown, d, max_dist)

        q = (cfg.z_hit * np.exp(-(d * d) / (2.0 * sigma * sigma)) +
             cfg.z_rand / scan.range_max)
        q = np.maximum(q, 1e-300)

        w = np.exp(np.sum(np.log(q), axis=1))
        return w

File: src/amr_localization/src/sensor_model.cpp
// SPDX-License-Identifier: Apache-2.0
#include "amr_localization/sensor_model.hpp"

#include <algorithm>
#include <cmath>
#include <cstdint>

#include "amr_core/geometry.hpp"

namespace amr_localization {

LikelihoodField::LikelihoodField(const amr_core::OccupancyGrid& grid,
                                 const amr_core::LikelihoodConfig& cfg)
    : grid_(grid), cfg_(cfg) {
  const std::size_t n =
      static_cast<std::size_t>(grid_.rows) * static_cast<std::size_t>(grid_.cols);
  // occupied iff data >= 65 (matches Python `grid.data >= 65`).
  std::vector<uint8_t> occupied(n);
  unknown_.assign(n, 0);
  for (std::size_t i = 0; i < n; ++i) {
    occupied[i] = (grid_.data[i] >= 65) ? 1u : 0u;
    // Cells unknown in the source map (data < 0) contribute no hit likelihood.
    unknown_[i] = (grid_.data[i] < 0) ? 1 : 0;
  }
  field_ = amr_core::distance_field(occupied, grid_.rows, grid_.cols,
                                    grid_.resolution, cfg_.max_dist);
}

std::vector<double> LikelihoodField::weigh(
    const std::vector<amr_core::Pose2D>& particles,
    const amr_core::LaserScan& scan) const {
  const std::size_t n = particles.size();
  const double res = grid_.resolution;
  const double max_dist = cfg_.max_dist;
  const double sigma = cfg_.sigma_hit;

  const std::vector<char> valid = scan.valid_mask();
  const int sub = std::max(cfg_.beam_subsample, 1);
  const std::vector<double> angles = scan.angles();
  const std::vector<double>& ranges = scan.ranges;

  // Subsampled valid beams: take valid beams, then stride by `sub` over that
  // filtered sequence (matches Python angles[valid][::sub]).
  std::vector<double> a;
  std::vector<double> r;
  std::size_t valid_count = 0;
  for (std::size_t i = 0; i < ranges.size(); ++i) {
    if (!valid[i]) {
      continue;
    }
    if (valid_count % static_cast<std::size_t>(sub) == 0) {
      a.push_back(angles[i]);
      r.push_back(ranges[i]);
    }
    ++valid_count;
  }

  if (a.empty()) {
    return std::vector<double>(n, 1.0);
  }

  const std::size_t b = a.size();
  // Beam endpoints in the robot frame.
  std::vector<double> bx(b);
  std::vector<double> by(b);
  for (std::size_t j = 0; j < b; ++j) {
    bx[j] = r[j] * std::cos(a[j]);
    by[j] = r[j] * std::sin(a[j]);
  }

  const double two_sigma2 = 2.0 * sigma * sigma;
  const double z_rand_term = cfg_.z_rand / scan.range_max;

  std::vector<double> w(n);
  for (std::size_t p = 0; p < n; ++p) {
    const double px = particles[p].x;
    const double py = particles[p].y;
    const double cos_t = std::cos(particles[p].theta);
    const double sin_t = std::sin(particles[p].theta);

    double log_sum = 0.0;
    for (std::size_t j = 0; j < b; ++j) {
      const double wx = px + cos_t * bx[j] - sin_t * by[j];
      const double wy = py + sin_t * bx[j] + cos_t * by[j];

      const long col = static_cast<long>(
          std::floor((wx - grid_.origin_x) / res));
      const long row = static_cast<long>(
          std::floor((wy - grid_.origin_y) / res));

      const bool in_bounds = (row >= 0 && row < grid_.rows && col >= 0 &&
                              col < grid_.cols);
      long rc = row;
      long cc = col;
      if (rc < 0) rc = 0;
      if (rc > grid_.rows - 1) rc = grid_.rows - 1;
      if (cc < 0) cc = 0;
      if (cc > grid_.cols - 1) cc = grid_.cols - 1;

      const std::size_t flat =
          static_cast<std::size_t>(rc) * grid_.cols + static_cast<std::size_t>(cc);
      double d;
      if (in_bounds && !unknown_[flat]) {
        d = static_cast<double>(field_[flat]);
      } else {
        // Out-of-bounds or unknown cells contribute the max distance.
        d = max_dist;
      }

      double q = cfg_.z_hit * std::exp(-(d * d) / two_sigma2) + z_rand_term;
      if (q < 1e-300) {
        q = 1e-300;
      }
      log_sum += std::log(q);
    }
    w[p] = std::exp(log_sum);
  }
  return w;
}

}  // namespace amr_localization
